<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Yuichi Shiraishi" />

<meta name="date" content="2015-01-30" />

<title>Introduction to pmsignature</title>




<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title">Introduction to pmsignature</h1>
<h4 class="author"><em>Yuichi Shiraishi</em></h4>
<h4 class="date"><em>2015-01-30</em></h4>
</div>


<p>The R package <em>pmsignature</em> is developed for efficiently extracting characteristic mutation patterns (mutation signatures) from the set of mutations collected typically from cancer genome sequencing data.</p>
<p>For extracting mutation signatures, principal component analysis or nonnegative matrix factorization have been popular. Compared to these existing approaches, the <code>pmsignature</code> has following advantages:</p>
<ol style="list-style-type: decimal">
<li><em>pmsignature</em> can perform robust estimation of mutation signatures even taking account for many mutation features such as two bases 5’ and 3’ to the mutated sites.</li>
<li><em>pmsignature</em> provides intuitively interetable visualization of mutation signatures, which is reminicent of sequencing logos.</li>
</ol>
<p>Currently, <em>pmsignature</em> can only accept tab delimited text files with specialized format. We will improve the program so that it can accept VCF format files.</p>
<div id="paper" class="section level2">
<h2>Paper</h2>
<blockquote>
<p>Shiraishi et al. Extraction of Latent Probabilistic Mutational Signature in Cancer Genomes, submitted.</p>
</blockquote>
</div>
<div id="input-data" class="section level2">
<h2>Input data</h2>
<p>For input data, we need mutation feature data for each sample and mutation. Here, <strong>mutation features</strong> are elements used for categorize the mutations such as:</p>
<ul>
<li>6 substitutions (C&gt;A, C&gt;G, C&gt;T, T&gt;A, T&gt;C and T&gt;G)</li>
<li>5’ and 3’ flanking bases (A, C, G and T)</li>
<li>transcription direction.</li>
</ul>
<p>Currently, <em>pmsignature</em> can accept following two formats of tab-delimited text file.</p>
<div id="mutation-position-format" class="section level3">
<h3>Mutation Position Format</h3>
<table>
<tbody>
<tr class="odd">
<td align="left">sample1</td>
<td align="left">chr1</td>
<td align="left">100</td>
<td align="left">A</td>
<td align="left">C</td>
</tr>
<tr class="even">
<td align="left">sample1</td>
<td align="left">chr1</td>
<td align="left">200</td>
<td align="left">A</td>
<td align="left">T</td>
</tr>
<tr class="odd">
<td align="left">sample1</td>
<td align="left">chr2</td>
<td align="left">100</td>
<td align="left">G</td>
<td align="left">T</td>
</tr>
<tr class="even">
<td align="left">sample2</td>
<td align="left">chr1</td>
<td align="left">300</td>
<td align="left">T</td>
<td align="left">C</td>
</tr>
<tr class="odd">
<td align="left">sample3</td>
<td align="left">chr3</td>
<td align="left">400</td>
<td align="left">T</td>
<td align="left">C</td>
</tr>
</tbody>
</table>
<ul>
<li>The 1st column shows the name of samples</li>
<li>The 2nd column shows the name of chromosome</li>
<li>The 3rd column shows the coordinate in the chromosome</li>
<li>The 4th column shows the reference base (A, C, G, or T).</li>
<li>The 5th colum shows the alternate base (A, C, G, or T).</li>
</ul>
</div>
<div id="mutation-feature-vector-format" class="section level3">
<h3>Mutation Feature Vector Format</h3>
<table>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="left">4</td>
<td align="left">4</td>
<td align="left">4</td>
<td align="left">3</td>
<td align="left">3</td>
<td align="left">2</td>
</tr>
<tr class="even">
<td align="left">2</td>
<td align="left">4</td>
<td align="left">3</td>
<td align="left">3</td>
<td align="left">1</td>
<td align="left">1</td>
<td align="left">2</td>
</tr>
<tr class="odd">
<td align="left">3</td>
<td align="left">4</td>
<td align="left">4</td>
<td align="left">3</td>
<td align="left">2</td>
<td align="left">2</td>
<td align="left">2</td>
</tr>
<tr class="even">
<td align="left">4</td>
<td align="left">3</td>
<td align="left">3</td>
<td align="left">2</td>
<td align="left">3</td>
<td align="left">3</td>
<td align="left">1</td>
</tr>
<tr class="odd">
<td align="left">5</td>
<td align="left">3</td>
<td align="left">4</td>
<td align="left">2</td>
<td align="left">4</td>
<td align="left">4</td>
<td align="left">2</td>
</tr>
<tr class="even">
<td align="left">6</td>
<td align="left">4</td>
<td align="left">1</td>
<td align="left">4</td>
<td align="left">2</td>
<td align="left">1</td>
<td align="left">2</td>
</tr>
<tr class="odd">
<td align="left">3</td>
<td align="left">2</td>
<td align="left">1</td>
<td align="left">1</td>
<td align="left">1</td>
<td align="left">1</td>
<td align="left">2</td>
</tr>
<tr class="even">
<td align="left">7</td>
<td align="left">4</td>
<td align="left">2</td>
<td align="left">2</td>
<td align="left">4</td>
<td align="left">3</td>
<td align="left">2</td>
</tr>
</tbody>
</table>
<ul>
<li>The 1st column shows the name of samples</li>
<li>From the 2nd to the last column show the value of mutation features.</li>
<li>In this example, substitution patterns (1 to 6 values, C&gt;A, C&gt;G, C&gt;T, T&gt;A, T&gt;C and T&gt;G), two 5’ and 3’ flanking bases (1 to 4 values for each 4 site, A, C, G and T) and transcription direction (1 to 2 values, + and -) are considered.</li>
</ul>
</div>
</div>
<div id="workflow" class="section level2">
<h2>Workflow</h2>
<div id="install-the-package" class="section level3">
<h3>Install the package</h3>
<p>First, the R packages <em>VariantAnnotation</em> and <em>BSgenome.Hsapiens.UCSC.hg19</em>, which <em>pmsignature</em> depends has to be installed. Also, <em>devtools</em> may be necessary for ease of installation.</p>
<pre><code>source(&quot;http://bioconductor.org/biocLite.R&quot;)
biocLite(c(&quot;VariantAnnotation&quot;, &quot;BSgenome.Hsapiens.UCSC.hg19&quot;))
install.packages(&quot;devtools&quot;)</code></pre>
<p>The easiest way for installing <em>pmsignature</em> is to use the package <em>devtools</em>:</p>
<pre><code>library(devtools)
devtools::install_github(&quot;friend1ws/pmsignature&quot;)</code></pre>
</div>
<div id="prepare-input-data" class="section level3">
<h3>Prepare input data</h3>
<p>First, create the input data from your mutation data.</p>
<p>After installing <em>pmsignature</em>, you can find the above example file at the directory where pmsignature is installed.</p>
<ul>
<li><p>Mutation Position Format</p>
<pre><code>inputFile &lt;- system.file(&quot;extdata/Nik_Zainal_2012.mutationPositionFormat.txt&quot;, package=&quot;pmsignature&quot;);
print(inputFile);</code></pre></li>
<li><p>Mutation Feature Vector Format</p>
<pre><code>inputFile &lt;- system.file(&quot;extdata/Hoang_MFVF.txt&quot;, package=&quot;pmsignature&quot;);
print(inputFile);</code></pre></li>
</ul>
</div>
<div id="read-input-data" class="section level3">
<h3>Read input data</h3>
<p>Type the following commands (<em>inputFile</em> is the path of the data you want to analyze):</p>
<ul>
<li><p>Mutation Position Format</p>
<pre><code>G &lt;- readMPFile(inputFile, numBases = 5);</code></pre>
<p>Here, <em>inputFile</em> is the path for the input file. <em>numBases</em> is the number of flanking bases to consider including the central base (if you want to consider two 5’ and 3’ bases, then set 5). You can rformat the data as the full model by typing</p>
<pre><code>G &lt;- readMPFile(inputFile, numBases = 5, type = &quot;full&quot;);</code></pre>
<p>Also, you can add transcription direction information by typing</p>
<pre><code>G &lt;- readMPFile(inputFile, numBases = 5, trDir = TRUE);</code></pre></li>
<li><p>Mutation Feature Vector Format</p>
<pre><code>G &lt;- readMFVFile(inputFile, numBases = 5);</code></pre></li>
</ul>
</div>
<div id="estimate-the-parameters" class="section level3">
<h3>Estimate the parameters</h3>
<p>When you want to set the number of mutation signature as 3, type the following command:</p>
<pre><code>Param &lt;- getPMSignature(G, K = 3);</code></pre>
<p>If you want to add the background signature, then after obtaining the background probability, perform the estimation. Currently, we only provide the background data for the “independent” and “full” model with 3 and 5 flanking bases.</p>
<pre><code>BG_prob &lt;- readBGFile(G);
Param &lt;- getPMSignature(G, K = 3, BG = BG_prob);</code></pre>
<p>In default, we repeat the estimation 10 times by changing the initial value, and select the parameter with maximum likelihood. If you want to changet the repeat number, then</p>
<pre><code>Param &lt;- getPMSignature(G, K = 3, numInit=20);</code></pre>
</div>
<div id="visualing-the-mutation-signatures-and-memberships" class="section level3">
<h3>Visualing the mutation signatures and memberships</h3>
<p>You can check the mutation signature by typing</p>
<pre><code>visPMSignature(Param, 1)
visPMSignature(Param, 2)
visPMSignature(Param, 3)</code></pre>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
